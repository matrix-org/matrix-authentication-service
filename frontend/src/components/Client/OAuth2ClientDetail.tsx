// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { H3 } from "@vector-im/compound-web";
import { useTranslation } from "react-i18next";

import { FragmentType, useFragment } from "../../gql";
import { graphql } from "../../gql/gql";
import BlockList from "../BlockList/BlockList";
import ExternalLink from "../ExternalLink/ExternalLink";
import ClientAvatar from "../Session/ClientAvatar";
import SessionDetails from "../SessionDetail/SessionDetails";

import styles from "./OAuth2ClientDetail.module.css";

export const OAUTH2_CLIENT_FRAGMENT = graphql(/* GraphQL */ `
  fragment OAuth2Client_detail on Oauth2Client {
    id
    clientId
    clientName
    clientUri
    logoUri
    tosUri
    policyUri
    redirectUris
  }
`);

type Props = {
  client: FragmentType<typeof OAUTH2_CLIENT_FRAGMENT>;
};

const FriendlyExternalLink: React.FC<{ uri?: string }> = ({ uri }) => {
  if (!uri) {
    return null;
  }
  const url = new URL(uri);
  const friendlyUrl = url.host + url.pathname;

  return <ExternalLink href={uri}>{friendlyUrl}</ExternalLink>;
};

const OAuth2ClientDetail: React.FC<Props> = ({ client }) => {
  const data = useFragment(OAUTH2_CLIENT_FRAGMENT, client);
  const { t } = useTranslation();

  const details = [
    { label: t("frontend.oauth2_client_detail.name"), value: data.clientName },
    {
      label: t("frontend.oauth2_client_detail.terms"),
      value: data.tosUri && <FriendlyExternalLink uri={data.tosUri} />,
    },
    {
      label: t("frontend.oauth2_client_detail.policy"),
      value: data.policyUri && <FriendlyExternalLink uri={data.policyUri} />,
    },
  ].filter(({ value }) => !!value);

  return (
    <BlockList>
      <header className={styles.header}>
        <ClientAvatar
          logoUri={data.logoUri || undefined}
          name={data.clientName || data.clientId}
          size="1.5rem"
        />
        <H3>{data.clientName}</H3>
      </header>
      <SessionDetails
        title={t("frontend.oauth2_client_detail.details_title")}
        details={details}
      />
    </BlockList>
  );
};

export default OAuth2ClientDetail;
