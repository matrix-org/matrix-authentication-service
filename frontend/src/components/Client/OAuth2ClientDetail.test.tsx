// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// @vitest-environment happy-dom

import { render, cleanup } from "@testing-library/react";
import { describe, expect, it, afterEach } from "vitest";

import { makeFragmentData } from "../../gql/fragment-masking";

import OAuth2ClientDetail, {
  OAUTH2_CLIENT_FRAGMENT,
} from "./OAuth2ClientDetail";

describe("<OAuth2ClientDetail>", () => {
  const baseClient = {
    id: "test-id",
    clientId: "client-id",
    clientName: "Test Client",
    clientUri: "https://client.org/logo.png",
    tosUri: "https://client.org/tos",
    policyUri: "https://client.org/policy",
    redirectUris: ["https://client.org/"],
  };
  afterEach(cleanup);

  it("renders client details", () => {
    const data = makeFragmentData(baseClient, OAUTH2_CLIENT_FRAGMENT);

    const { container } = render(<OAuth2ClientDetail client={data} />);

    expect(container).toMatchSnapshot();
  });

  it("does not render terms of service when falsy", () => {
    const data = makeFragmentData(
      {
        ...baseClient,
        tosUri: undefined,
      },
      OAUTH2_CLIENT_FRAGMENT,
    );

    const { queryByText } = render(<OAuth2ClientDetail client={data} />);

    expect(queryByText("Terms of service")).toBeFalsy();
  });

  it("does not render logo when logoUri is falsy", () => {
    const data = makeFragmentData(
      {
        ...baseClient,
        logoUri: undefined,
      },
      OAUTH2_CLIENT_FRAGMENT,
    );

    const { queryByAltText } = render(<OAuth2ClientDetail client={data} />);

    expect(queryByAltText(baseClient.clientName)).toBeFalsy();
  });
});
