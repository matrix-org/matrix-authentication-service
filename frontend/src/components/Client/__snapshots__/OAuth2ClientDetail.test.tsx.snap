// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<OAuth2ClientDetail> > renders client details 1`] = `
<div>
  <div
    class="_blockList_f8cc7f"
  >
    <header
      class="_header_477219"
    >
      <h3
        class="_font-heading-md-semibold_1jx6b_121"
      >
        Test Client
      </h3>
    </header>
    <div
      class="_block_17898c"
    >
      <h6
        class="_font-body-md-semibold_1jx6b_64"
      >
        Client
      </h6>
      <ul
        class="_list_040867"
      >
        <li
          class="_detailRow_040867"
        >
          <p
            class="_font-body-sm-semibold_1jx6b_45 _detailLabel_040867"
          >
            Name
          </p>
          <p
            class="_font-body-sm-regular_1jx6b_40 _detailValue_040867"
          >
            Test Client
          </p>
        </li>
        <li
          class="_detailRow_040867"
        >
          <p
            class="_font-body-sm-semibold_1jx6b_45 _detailLabel_040867"
          >
            Client ID
          </p>
          <p
            class="_font-body-sm-regular_1jx6b_40 _detailValue_040867"
          >
            <code>
              client-id
            </code>
          </p>
        </li>
        <li
          class="_detailRow_040867"
        >
          <p
            class="_font-body-sm-semibold_1jx6b_45 _detailLabel_040867"
          >
            Terms of service
          </p>
          <p
            class="_font-body-sm-regular_1jx6b_40 _detailValue_040867"
          >
            <a
              class="_link_eqn56_17 _externalLink_a97355"
              data-kind="primary"
              href="https://client.org/tos"
              rel="noreferrer noopener"
              target="_blank"
            >
              client.org/tos
            </a>
          </p>
        </li>
        <li
          class="_detailRow_040867"
        >
          <p
            class="_font-body-sm-semibold_1jx6b_45 _detailLabel_040867"
          >
            Policy
          </p>
          <p
            class="_font-body-sm-regular_1jx6b_40 _detailValue_040867"
          >
            <a
              class="_link_eqn56_17 _externalLink_a97355"
              data-kind="primary"
              href="https://client.org/policy"
              rel="noreferrer noopener"
              target="_blank"
            >
              client.org/policy
            </a>
          </p>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
